dataset.py

```diff
@@ -0,0 +1,161 @@
+                                         0.5)
+
+def gen_tabular_dataset_with_coords(monthly: bool, samples_per_site: int, 
+  sample_site_coordinates: list, sample_radius: float) -> pd.DataFrame:
+  features = [raster.relative_humidity_geotiff(), 
+    raster.temperature_geotiff(), 
+    raster.vapor_pressure_deficit_geotiff(), 
+    raster.atmosphere_isoscape_geotiff(), 
+    raster.cellulose_isoscape_geotiff()]
+  image_feature_names = ["rh", "temp", "vpd", "atmosphere_oxygen_ratio", "cellulose_oxygen_ratio"]
+  feature_names = ["lat", "lon", "month_of_year"] + image_feature_names
+  rs = RandomState(MT19937(SeedSequence(42)))
+
+  feature_values = {}
+  for name in feature_names:
+    feature_values[name] = []
+
+  for coord in tqdm(sample_site_coordinates):
+    month_start = 0 if monthly else -1
+    month_end = 12 if monthly else 0
+    for month in range(month_start, month_end):
+      samples_collected = 0
+      while samples_collected < samples_per_site:
+        row = {}
+        sample_x, sample_y = 2*(rs.rand(2) - 0.5) * sample_radius
+        sample_x += coord[0]
+        sample_y += coord[1]
+
+        try:
+          for feature, feature_name in zip(features, image_feature_names):
+            row[feature_name] = raster.get_data_at_coords(
+              feature, sample_x, sample_y, month)
+          row["month_of_year"] = month
+          row["lon"] = sample_x
+          row["lat"] = sample_y
+          samples_collected += 1
+
+        except ValueError as e:
+          # masked and out-of-bounds coordinates
+          # note that if sample_radius is zero, we know that looping again
+          # will still result in a ValueError. In this case, we trigger the
+          # loop to finish by incrementing the end condition (samples_collected)
+          # In that case, we end up ignoring the point.
+          # This can also happen with a small positive radius, but one whose
+          # coords return all masked/out of bounds, but this is less likely.
+          if sample_radius == 0:
+            samples_collected += 1
+          continue
+        for key, value in row.items():
+          feature_values[key].append(value)
+
+  samples = pd.DataFrame(feature_values)
+
+  if not monthly:
+    samples.drop("month_of_year", axis=1, inplace=True)
+
+  return samples
+
+"""This method loads reference sample data from a CSV, determines the unique
+locations in that CSV and then calls the existing method
+gen_tabular_dataset_with_coords to enrich those locations with features for
+"rh", "temp", "vpd", and  "atmosphere_oxygen_ratio".
+Finally, it joins the CSV provided d18O means values from the CSV with the
+per-loc features"""
+def load_sample_data(reference_csv_filename: str) -> pd.DataFrame:
+  df = pd.read_csv(raster.get_sample_db_path(reference_csv_filename),
+   encoding="ISO-8859-1", sep=',')
+  df = df[['Code', 'lat', 'long', 'd18O_cel']]
+  df = df[df['d18O_cel'].notna()]
```

Comment:
I think C and N are likely missing because we were focused on O.
However, we will be moving to Luis' code for calculating the training sets, which should trim nans.  Are you ok with not fixing this until then?

Reply:
Sure, we can postpone